test/e2e/scheduling/predicates.go

```diff
@@ -724,6 +727,231 @@ var _ = SIGDescribe("SchedulerPredicates [Serial]", func() {
+			affinity := &v1.Affinity{
+				PodAntiAffinity: &v1.PodAntiAffinity{
+					RequiredDuringSchedulingIgnoredDuringExecution: []v1.PodAffinityTerm{
+						{
+							LabelSelector: &metav1.LabelSelector{
+								MatchExpressions: []metav1.LabelSelectorRequirement{
+									{
+										Key:      "security",
+										Operator: metav1.LabelSelectorOpIn,
+										Values:   []string{"S1"},
+									},
+								},
+							},
+							TopologyKey:    k,
+							Namespaces:     []string{ns},
+							MatchLabelKeys: []string{"pod-template-hash"},
+						},
+					},
+				},
+			}
+
+			intstrPtr := func(i int) *intstr.IntOrString {
+				num := intstr.FromInt(i)
+				return &num
+			}
+			deploymentPodLabels := map[string]string{"test": "matchLabelKey1", "security": "S1"}
+			deploymentName := "test-matchlabelkey-deployment"
+			deploymentReplicas := int32(numNode)
+			deploymentImage := imageutils.GetPauseImageName()
+			deploymentStrategyType := appsv1.RollingUpdateDeploymentStrategyType
+			newDeployment := e2edeployment.NewDeployment(deploymentName, deploymentReplicas, deploymentPodLabels, "pause", deploymentImage, deploymentStrategyType)
+			newDeployment.Spec.Strategy.RollingUpdate = &appsv1.RollingUpdateDeployment{
+				MaxUnavailable: intstrPtr(0),
+				MaxSurge:       intstrPtr(numNode),
+			}
+			newDeployment.Spec.Template.Spec.Affinity = affinity
+
+			d, err := cs.AppsV1().Deployments(ns).Create(ctx, newDeployment, metav1.CreateOptions{})
+			framework.ExpectNoError(err)
+
+			framework.Logf("Make sure deployment %q is complete", deploymentName)
+			pollInterval := 1 * time.Second
+			err = testutils.WaitForDeploymentCompleteAndCheckRolling(cs, d, framework.Logf, pollInterval, wait.ForeverTestTimeout)
+			framework.ExpectNoError(err)
+
+			framework.Logf("Make sure all Pods in deployment %q is scheduled", deploymentName)
+			err = wait.Poll(pollInterval, wait.ForeverTestTimeout, deploymentReplicasScheduled(ctx, cs, ns, deploymentPodLabels))
+			framework.ExpectNoError(err)
+
+			framework.Logf("Make sure all Pods in deployment %q is scheduled on different Node", deploymentName)
+			nodenames, err := deploymentReplicasNodeNames(ctx, cs, ns, deploymentPodLabels)
+			framework.ExpectNoError(err)
+			nodenameset := sets.Set[string]{}
+			for _, nodename := range nodenames {
+				nodenameset.Insert(nodename)
+			}
+			framework.ExpectEqual(len(nodenames), nodenameset.Len())
+
+			framework.Logf("update deployment %q to run rolling upgrade", deploymentName)
+			// change label to kick the rolling upgrade.
+			d.Spec.Template.Labels["test"] = "matchLabelKey2"
+			d, err = cs.AppsV1().Deployments(ns).Update(ctx, newDeployment, metav1.UpdateOptions{})
+			framework.ExpectNoError(err)
+
+			framework.Logf("Make sure updated deployment %q is complete", deploymentName)
+			err = testutils.WaitForDeploymentCompleteAndCheckRolling(cs, d, framework.Logf, pollInterval, wait.ForeverTestTimeout)
```

Comment:
I assume this function, along with deploymentReplicasNodeNames() can guarantee we are testing the **_new_** replicas, right? Just want to avoid edge case like the rolling upgrade is not happening at all, and it returns ok after checking **_old_** replicas.

Reply:
Yes, correct. 
This `WaitForDeploymentCompleteAndCheckRolling` makes sure that the new replicaset is created, and all Pods in new replicaset are created (= the rolling upgrade is completed). And, `deploymentReplicasScheduled` makes sure that all new replicas are scheduled.